eksrollup/cli.py

```diff
@@ -11,32 +11,58 @@
-                    cluster_healthy = True
+    nodes_ready = False
+    retry_count = 0
+
+    while retry_count < cluster_health_retry:
+        retry_count += 1
+        if health_check_type == "asg":
+            logger.info(f'Waiting for {cluster_health_wait} seconds for ASG to scale before validating cluster health...')
+        else:
+            logger.info(f'Waiting for {cluster_health_wait} seconds before validating cluster health...')
+
+        time.sleep(cluster_health_wait)
+        desired_k8s_node_count = count_all_cluster_instances(cluster_name, predictive=predictive)
+
+        # check if asg has enough nodes first before checking instance health
+        if is_asg_scaled(asg_name, new_desired_asg_capacity):
+            # if asg is healthy start draining and terminating instances
+            if is_asg_healthy(asg_name):
+                # check if k8s nodes are all online and with the matches the desired amount of instances
+                if k8s_nodes_count(desired_k8s_node_count):
+                    # check k8s nodes are healthy
+                    if k8s_nodes_ready():
+                        nodes_ready = True
+                        cluster_healthy = True
+                        break
                 else:
-                    logger.info('Validation failed for cluster. Expected node count reached but nodes are not healthy.')
-            else:
-                nodes = get_k8s_nodes()
-                logger.info('Current k8s node count is {}'.format(len(nodes)))
-                logger.info('Validation failed for cluster. Current node count {} Expected node count {}.'.format(
-                    len(nodes),
-                    desired_k8s_node_count))
+                    node_count_mismatch = True
         else:
-            logger.info(
-                'Validation failed for asg {}.'
-                'Instances not healthy'.format(asg_name))
+            logger.info('Validation failed for asg {}. Not enough instances online.'.format(asg_name))
+
+    if cluster_healthy:
+        logger.info('Cluster validation passed. Proceeding with node draining and termination...')
+    elif not nodes_ready:
+        logger.info('Validation failed for cluster. Expected node count reached but nodes are not healthy.')
+    elif node_count_mismatch:
+        nodes = get_k8s_nodes()
+        logger.info('Current k8s node count is {}'.format(len(nodes)))
+        logger.info('Validation failed for cluster. Current node count {} Expected node count {}.'.format(
+            len(nodes),
+            desired_k8s_node_count))
     else:
-        logger.info('Validation failed for asg {}. Not enough instances online.'.format(asg_name))
+        logger.info(
+            'Validation failed for asg {}.'
+            'Instances not healthy'.format(asg_name))
+
+    if not cluster_healthy:
+        logger.info('Exiting since ASG healthcheck failed')
+        raise Exception('ASG healthcheck failed')
+
     return cluster_healthy
```

Comment:
Nitpick: this return value is unnecessary now, can remove for clarity?

Reply:
Yes I removed it now.